### ISSUE TITLE
No error message when distributed optimizer is used for gluon trainer

### ISSUE THREAD

**Issue by eric-haibin-lin** [horovod/horovod#1657]
```
optim = hvd.DistributedOptimizer(optim)
trainer = gluon.Trainer(param_dict,
                                     optim,
                                     update_on_kvstore=False)
```
I found that there is no error message if I use distributed optimizer inside gluon trainer. This is not recommended way, and deserves an error message. 

@apeforest

**Comment by apeforest** [comment]
Acknowledged issue. I will add an error message when users mix DistributedOptimizer with regular Gluon Trainer.

**Comment by karan6181** [comment]
Agree with your point. So I was looking into class `DistributedOptimizer` and `DistributedTrainer`. The `DistributedTrainer` checks if the optimizer is of instance of `DistributedOptimizer` or `mx.optimizer.Optimizer` and adjust accordingly. Since the `gluon.Trainer` takes optimizer as a part of a parameter, how are we gonna put a checking condition in horovod since implementation of `gluon.Trainer` is in MXNet? We don't want to modify the implementation of `gluon.Trainer` and put a checking condition there. One way is to put the warning message in class `DistributedOptimizer` to notify user to not use `gluon.Trainer` if using imperative API and use only for Module API.

**Comment by eric-haibin-lin** [comment]
Good point. It might be hard to check this. We don't want to add horovod specific code in gluon Trainer. I am not sure about putting warning message in distributed optimizer - if the user is using the API correctly and see this warning, they will get confused
